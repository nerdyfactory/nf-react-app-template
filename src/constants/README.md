## constants
constants such as static strings or CODEs
