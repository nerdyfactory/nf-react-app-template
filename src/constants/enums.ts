export enum MUIColorEnums {
  primary = 'primary',
  inherit = 'inherit',
  secondary = 'secondary',
  default = 'default',
}

export enum MUIButtonVariantEnums {
  contained = 'contained',
  outlined = 'outlined',
  text = 'text',
}
