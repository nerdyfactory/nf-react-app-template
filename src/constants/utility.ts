export const USER_TOKEN = `NF_USER_TOKEN`;

export const IS_DEVELOPMENT = process.env.NODE_ENV === 'development';

export const ENDPOINTS = {
  GRAPHQL_URI: IS_DEVELOPMENT ? 'http://localhost:8082/api/graphql' : '/api/graphql',
  BASE_URL: IS_DEVELOPMENT ? 'http://localhost:8082/api/' : '/api/',
};

export const ERROR_MESSAGES = {
  REQUIRED_FIELD: `필수 필드`,
  INVALID_EMAIL: `이메일이 유효하지 않습니다`,
  PASSWORD_INVALID: `암호가 유효하지 않습니다`,
  PASSWORD_NOT_MATCHING: `비밀번호가 일치하지 않습니다`,
  PHONE_INVALID: `잘못된 전화 번호`,
};

export const PASSWORD_REGEX = /^(?=.*[a-zA-Z])(?=.*[0-9])/;
export const PHONE_REGEX = /^\d{3}-\d{3,4}-\d{4}$/;

export const USER_ROLES = {
  MANAGER: 'Manager',
  ADMIN: 'Admin',
};
