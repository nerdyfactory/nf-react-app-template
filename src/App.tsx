import React, { useState, useReducer, useEffect } from 'react';
import { ThemeProvider } from '@material-ui/styles/';
import { BrowserRouter } from 'react-router-dom';
import { Toaster } from 'react-hot-toast';

import { PrivateNav, PublicNav } from 'navigations';
import useBootstrap from 'hooks/useBootstrap';
import { AppContext, appReducer, INITIAL_APP_STATE } from 'contexts/AppContext';
import { AuthActionTypes, AuthContext, authReducer, INITIAL_AUTH_STATE } from 'contexts';
import { Loading } from 'components';
import './App.css';
import { USER_TOKEN } from 'constants/utility';
import { getIntialState } from 'services/utils/persist-state';

function App() {
  const [authCtx, authDispatcher] = useReducer(authReducer, INITIAL_AUTH_STATE);
  const [appCtx, appDispatcher] = useReducer(appReducer, INITIAL_APP_STATE);
  const [theme] = useBootstrap();
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    const token = getIntialState(USER_TOKEN);
    if (token) {
      authDispatcher({ type: AuthActionTypes.SET_TOKEN, payload: token });
    }
    setIsLoaded(true);
  }, []);

  if (!isLoaded) return <Loading />;

  return (
    <ThemeProvider theme={theme}>
      <BrowserRouter basename="/">
        <AuthContext.Provider value={{ ...authCtx, dispatch: authDispatcher }}>
          <AppContext.Provider value={{ ...appCtx, dispatch: appDispatcher }}>
            {authCtx.token ? <PrivateNav /> : <PublicNav />}
            <Toaster
              position="top-right"
              toastOptions={{
                style: {
                  margin: '40px',
                  background: '#363636',
                  color: '#fff',
                  zIndex: 99999999,
                },
              }}
            />
          </AppContext.Provider>
        </AuthContext.Provider>
      </BrowserRouter>
    </ThemeProvider>
  );
}

export default App;
