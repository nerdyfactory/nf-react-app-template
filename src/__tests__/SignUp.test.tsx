import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import 'jest-localstorage-mock';
import App from '../App';
import { ERROR_MESSAGES } from 'constants/utility';

jest.mock('../services/api');

const VALID_EMAIL = 'test@test.com';
const VALID_PASSWORD = 'test1234';

describe('Sign Up', () => {
  it('renders Sign Up', async () => {
    const res = render(<App />);
    fireEvent.click(screen.getByText('Create account'));
    expect(res.container).toMatchSnapshot();
  });

  it('catches empty fields', async () => {
    render(<App />);
    fireEvent.input(screen.getByPlaceholderText('Email'), {
      target: {
        value: ``,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Password'), {
      target: {
        value: ``,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Confirm Password'), {
      target: {
        value: ``,
      },
    });
    fireEvent.click(screen.getByText('Sign Up'));
    await waitFor(() => screen.getByText('Sign Up'));
    expect(screen.getAllByText(ERROR_MESSAGES.REQUIRED_FIELD).length).toBe(3);
  });

  it('catches invalid email', async () => {
    render(<App />);
    fireEvent.input(screen.getByPlaceholderText('Email'), {
      target: {
        value: `wrong_username`,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Password'), {
      target: {
        value: VALID_PASSWORD,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Confirm Password'), {
      target: {
        value: VALID_PASSWORD,
      },
    });
    fireEvent.click(screen.getByText('Sign Up'));
    await waitFor(() => screen.getByText('Sign Up'));
    expect(screen.getAllByRole('button')[0]).toHaveTextContent('Sign Up');
    expect(screen.getByText(ERROR_MESSAGES.INVALID_EMAIL)).toBeInTheDocument();
  });

  it('catches invalid password', async () => {
    render(<App />);
    fireEvent.input(screen.getByPlaceholderText('Email'), {
      target: {
        value: VALID_EMAIL,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Password'), {
      target: {
        value: `test`,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Confirm Password'), {
      target: {
        value: `test`,
      },
    });
    fireEvent.click(screen.getByText('Sign Up'));
    await waitFor(() => screen.getByText('Sign Up'));
    expect(screen.getAllByRole('button')[0]).toHaveTextContent('Sign Up');
    expect(screen.getByText(ERROR_MESSAGES.PASSWORD_INVALID)).toBeInTheDocument();
  });

  it('catches passwords not matching', async () => {
    render(<App />);
    fireEvent.input(screen.getByPlaceholderText('Email'), {
      target: {
        value: VALID_EMAIL,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Password'), {
      target: {
        value: `test1234`,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Confirm Password'), {
      target: {
        value: `test4321`,
      },
    });
    fireEvent.click(screen.getByText('Sign Up'));
    await waitFor(() => screen.getByText('Sign Up'));
    expect(screen.getAllByRole('button')[0]).toHaveTextContent('Sign Up');
    expect(screen.getByText(ERROR_MESSAGES.PASSWORD_NOT_MATCHING)).toBeInTheDocument();
  });

  it('moves to login after sign up', async () => {
    render(<App />);
    fireEvent.input(screen.getByPlaceholderText('Email'), {
      target: {
        value: VALID_EMAIL,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Password'), {
      target: {
        value: VALID_PASSWORD,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Confirm Password'), {
      target: {
        value: VALID_PASSWORD,
      },
    });
    fireEvent.click(screen.getByText('Sign Up'));
    await waitFor(() => screen.getByText('Sign Up'));
    expect(screen.getAllByRole('button')[0]).toHaveTextContent('Login');
  });
});
