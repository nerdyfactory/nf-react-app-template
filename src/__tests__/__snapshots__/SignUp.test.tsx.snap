// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up renders Sign Up 1`] = `
<div>
  <div
    class="sc-dlfnbm iuHaXN"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz eHbxyQ"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="standard-basic"
          placeholder="User"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz eHbxyQ"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="standard-basic"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz eHbxyQ"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="standard-basic"
          placeholder="Confirm Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-bdfBwQ lfqaYk MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Sign Up
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text sc-bdfBwQ lfqaYk MuiButton-textPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Login
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    style="position: fixed; z-index: 9999;"
  />
</div>
`;
